### ISSUE TITLE
Intermittent Blob Unknown error during docker push operation

### ISSUE THREAD

**Issue by InderpreetSaini** [goharbor/harbor#16777]
Hi,
I am trying to deploy Harbor Helm Chart 2.5.0 in a Kubernetes RKE2 cluster with 3 master -3 worker nodes. But I am facing an intermittent issue when the docker push fails and Blob unknown error is thrown instead. I am deploying 3 replicas of each component including registry component and using NFS Storage as the persistent volume. The same setup was working fine for Harbor 2.4.1 helm chart.

There is no reverse proxy involved and I have already tried pruning the docker build server to remove all images to verify for any image/layer contamination, but the error still persists.

Also checked the redis/sentinel pod logs, but did not see any error there. I am deploying the Redis bitnami Helm chart. I am using only password authentication for Redis, not for Sentinel(last checked not supported by Harbor)

This is the error I am seeing in the registry pod logs.

`time="2022-04-28T15:54:13.782351264Z" level=error msg="error connecting to redis instance redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 service=registry version=v2.8.0
time="2022-04-28T15:54:13.782373709Z" level=error msg="redis: error connecting: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 redis.connect.duration=39.65µs service=registry version=v2.8.0
time="2022-04-28T15:54:13.782470824Z" level=error msg="error connecting to redis instance redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 service=registry version=v2.8.0
time="2022-04-28T15:54:13.782487541Z" level=error msg="redis: error connecting: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 redis.connect.duration=17.604µs service=registry version=v2.8.0
time="2022-04-28T15:54:13.782627537Z" level=error msg="error connecting to redis instance redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 service=registry version=v2.8.0
time="2022-04-28T15:54:13.782646986Z" level=error msg="redis: error connecting: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 redis.connect.duration=21.971µs service=registry version=v2.8.0
time="2022-04-28T15:54:13.782717552Z" level=error msg="error connecting to redis instance redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 service=registry version=v2.8.0
time="2022-04-28T15:54:13.782736848Z" level=error msg="redis: error connecting: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 redis.connect.duration=22.23µs service=registry version=v2.8.0
time="2022-04-28T15:54:13.782758974Z" level=error msg="error connecting to redis instance redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 service=registry version=v2.8.0
time="2022-04-28T15:54:13.78277085Z" level=error msg="redis: error connecting: dial tcp: address redis-node-0.redis-headless.redis.svc.cluster.local:26379,redis-node-1.redis-headless.redis.svc.cluster.local:26379,redis-node-2.redis-headless.redis.svc.cluster.local:26379: too many colons in address" go.version=go1.17.7 instance.id=9f611597-36b6-4191-bc8e-30519d6606d7 redis.connect.duration=16.438µs service=registry version=v2.8.0
time="2022-04-28T15:54:13.782844197Z" level=error msg="response completed with error" auth.user.name="harbor_registry_user" err.code="manifest blob unknown" err.detail=sha256:8c811b4aec35f259572d0f79207bc0678df4c736eeec50bc9fec37ed936a472a err.message="blob unknown to registry" go.version=go1.17.7 http.request.contenttype="application/vnd.docker.distribution.manifest.v2+json" http.request.host=[redacted] http.request.id=173baa35-28f6-4084-be4d-e7d636acf473 http.request.method=PUT http.request.remoteaddr=53.137.17.45 http.request.uri="/v2/test-private-project/busybox1/manifests/1.28" http.request.useragent="docker/20.10.12-ce go/go1.16.13 git-commit/459d0dfbbb51 kernel/5.3.18-59.24-default os/linux arch/amd64 UpstreamClient(Docker-Client/20.10.12-ce \(linux\))" http.response.contenttype="application/json; charset=utf-8" http.response.duration=76.391153ms http.response.status=400 http.response.written=166 vars.name="test-private-project/busybox1" vars.reference=1.28`

Could you please help with this error?

**Comment by alexanderdalloz** [comment]
After upgrading from Harbor 2.3.5 -> 2.4.2 -> 2.5.0 I too see the error about registry container complaining about a syntax error connecting to redis-sentinel. The very same configuration has worked before.

`Apr 29 16:19:06 node001 docker-compose[1086029]: registry       | time="2022-04-29T14:19:06.020714027Z" level=error msg="error connecting to redis instance 10.6.10.20:26379,10.6.10.21:26379,10.6.10.22:26379: dial tcp: address 10.6.10.20:26379,10.6.10.2:26379,10.6.10.22:26379: too many colons in address" go.version=go1.17.7 instance.id=82e7c154-c3a3-41e8-a469-c40c5aea8844 service=registry version=v2.8.0`

```
# grep "  host: " /opt/harbor/harbor.yml
  host: 10.6.10.20:26379,10.6.10.21:26379,10.6.10.22:26379
```

Thus redis caching completely fails?

**Comment by wy65701436** [comment]
are you using redis in sentinel mode? If so, it would be fixed in v2.5.1.

**Comment by InderpreetSaini** [comment]
> are you using redis in sentinel mode? If so, it would be fixed in v2.5.1.

Thanks @wy65701436. Our requirement is to have a HA Redis setup. Until 2.5.1 is released, could you please let us know if Redis cluster is supported by 2.5.0, so that we can implement HA if sentinel is not currently supported?

**Comment by zyyw** [comment]
The issue of using redis in sentinel mode in Harbor v2.5.0 is because there is a bug when we upgrade distribution/distribution (REGISTRY_SRC_TAG) to v2.8.0 in Harbor v2.5.0 (the redis.patch doesn't apply correctly).
It is fixed in this PR: https://github.com/goharbor/harbor/pull/16787 And it will be shipped in Harbor v2.5.1

The distribution/distribution (REGISTRY_SRC_TAG) is not upgrade in harbor v2.4.2, and the redis behavior of Harbor 2.4.2 should be consistent with Harbor v2.4.1. You can have a try using redis in sentinel mode in Harbor v2.4.2.
